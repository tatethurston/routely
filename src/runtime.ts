import { getQueryType } from "./utils";

/**
 * Escapes RegExp control characters from a path segment
 * https://stackoverflow.com/questions/4669692/valid-characters-for-directory-part-of-a-url-for-short-links
 */
export function escapeUrlPathRegExp(str: string): string {
  return str.replace(/[*+$()//]/g, "\\$&");
}

export function getPathRegex(path: string): RegExp {
  const re = escapeUrlPathRegExp(path)
    // optional catch all
    .replaceAll(/\\\/\[\[\.\.\.(.+?)\]\]/g, "(?<$1>.*)")
    // catch all
    .replaceAll(/\[\.\.\.(.+?)\]/g, "(?<$1>.+)")
    // dynamic
    .replaceAll(/\[(.+?)\]/g, "(?<$1>[^/]+)");

  return new RegExp(`^${re}/*$`);
}

type Match = (path: string) => {
  match: boolean;
  params: Record<string, string | string[]>;
};

export function getMatch(pathname: string): Match {
  const query = getQueryType(pathname);
  const re = getPathRegex(pathname);

  return (path: string) => {
    const match = re.exec(path);
    if (!match) {
      return { match: false, params: {} };
    }

    const params = Object.entries(query).reduce<
      Record<string, string | string[]>
    >((acc, [name, _type]) => {
      if (_type === "catch-all" || _type === "optional-catch-all") {
        // normalize matches from regex that may be:
        // '', '/', to []
        // and '/a', '/a/b/c/ to ['a'] and ['a', 'b', 'c']
        acc[name] = match.groups?.[name].split("/").filter(Boolean) ?? [];
      } else {
        acc[name] = match.groups?.[name] ?? "";
      }
      return acc;
    }, {});
    return { match: true, params };
  };
}
