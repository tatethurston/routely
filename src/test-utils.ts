import { routeFromFilepath } from "./utils";

export const filepaths = [
  "/404.ts",
  "/[foo].ts",
  "/[foo]/[bar]/[baz].ts",
  "/[foo]/bar/[baz].ts",
  "/[foo]/bar/[baz]/foo/[bar].ts",
  "/[foo]/baz.ts",
  "/_app.ts",
  "/_debug.ts",
  "/_debug/health-check.ts",
  "/_document.ts",
  "/_error.ts",
  "/_error/index.ts",
  "/api/[[...segments]].ts",
  "/api/[...segments].ts",
  "/api/bar.ts",
  "/foo/[slug].ts",
  "/index.ts",
  "/not-found.ts",
  "/settings/bars/[bar].ts",
  "/settings/bars/[bar]/baz.ts",
  "/settings/foo.ts",
  "/settings/index.ts",
];

export const routes = filepaths.map(routeFromFilepath);
