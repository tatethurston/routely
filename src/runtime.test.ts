import { router, getMatch, escapeUrlPathRegExp, getPathRegex } from "./runtime";
import { routes } from "./test-utils";

describe("escapeRegExp", () => {
  it.each(routes)("escapes %s", (path) =>
    expect(escapeUrlPathRegExp(path)).toMatchSnapshot()
  );
});

describe("getPathRegex", () => {
  it.each(routes)("escapes %s", (path) =>
    expect(getPathRegex(path)).toMatchSnapshot()
  );

  describe("static routes", () => {
    it("matches", () => {
      let route;

      route = "/";
      expect(getPathRegex(route).test(route)).toEqual(true);
      expect(getPathRegex(route).test("/foo")).toEqual(false);

      route = "/foo";
      expect(getPathRegex(route).test(route)).toEqual(true);
      expect(getPathRegex(route).test(route + "/")).toEqual(true);
      expect(getPathRegex(route).test("/")).toEqual(false);

      route = "/foo/bar";
      expect(getPathRegex(route).test(route)).toEqual(true);
      expect(getPathRegex(route).test(route + "/")).toEqual(true);
      expect(getPathRegex(route).test("/foo")).toEqual(false);
    });
  });

  describe("dynamic routes", () => {
    it("matches", () => {
      let route;

      route = "/[foo]";
      expect(getPathRegex(route).test("/foo")).toEqual(true);
      expect(getPathRegex(route).test("/bar")).toEqual(true);
      expect(getPathRegex(route).test("/bar/")).toEqual(true);
      expect(getPathRegex(route).test("/")).toEqual(false);
      expect(getPathRegex(route).test("/bar/foo")).toEqual(false);

      route = "/[foo]/[bar]";
      expect(getPathRegex(route).test("/foo/bar")).toEqual(true);
      expect(getPathRegex(route).test("/bar/baz")).toEqual(true);
      expect(getPathRegex(route).test("/bar/baz/")).toEqual(true);
      expect(getPathRegex(route).test("/foo")).toEqual(false);
      expect(getPathRegex(route).test("/foo/")).toEqual(false);
      expect(getPathRegex(route).test("/foo/bar/baz")).toEqual(false);

      route = "/[foo]/[bar]/[baz]";
      expect(getPathRegex(route).test("/foo/bar/baz")).toEqual(true);
      expect(getPathRegex(route).test("/bar/baz/foo")).toEqual(true);
      expect(getPathRegex(route).test("/bar/baz/foo/")).toEqual(true);
      expect(getPathRegex(route).test("/foo/bar")).toEqual(false);
      expect(getPathRegex(route).test("/foo/bar/baz/foo")).toEqual(false);

      route = "/foos/[foo]/bars/[bar]/bazs/[baz]";
      expect(getPathRegex(route).test("/foos/1/bars/2/bazs/3")).toEqual(true);
      expect(getPathRegex(route).test("/bar/baz/foo")).toEqual(false);
    });

    it("extracts query parameters", () => {
      let route;

      route = "/[foo]";
      expect(getPathRegex(route).exec("/bar")?.groups).toEqual({
        foo: "bar",
      });

      route = "/[foo]/[bar]";
      expect(getPathRegex(route).exec("/bar/baz")?.groups).toEqual({
        bar: "baz",
        foo: "bar",
      });

      route = "/[foo]/[bar]/[baz]";
      expect(getPathRegex(route).exec("/1/2/3")?.groups).toEqual({
        bar: "2",
        baz: "3",
        foo: "1",
      });

      route = "/foos/[foo]/bars/[bar]/bazs/[baz]";
      expect(getPathRegex(route).exec("/foos/1/bars/2/bazs/3")?.groups).toEqual(
        {
          bar: "2",
          baz: "3",
          foo: "1",
        }
      );
    });
  });

  describe("catch all routes", () => {
    it("matches", () => {
      const route = "/post/[...slug]";
      expect(getPathRegex(route).test("/post/a")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/b")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/b/c")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/b/c/")).toEqual(true);
      expect(getPathRegex(route).test("/post")).toEqual(false);
      expect(getPathRegex(route).test("/post/")).toEqual(false);
    });

    it("extracts query parameters", () => {
      const route = "/post/[...slug]";
      expect(getPathRegex(route).exec("/post/a")?.groups).toEqual({
        slug: "a",
      });
      expect(getPathRegex(route).exec("/post/a/")?.groups).toEqual({
        slug: "a/",
      });
      expect(getPathRegex(route).exec("/post/a/b")?.groups).toEqual({
        slug: "a/b",
      });
      expect(getPathRegex(route).exec("/post/a/b/c")?.groups).toEqual({
        slug: "a/b/c",
      });
      expect(getPathRegex(route).exec("/post/a/b/c/")?.groups).toEqual({
        slug: "a/b/c/",
      });
    });
  });

  describe("optional catch all routes", () => {
    it("matches", () => {
      const route = "/post/[[...slug]]";
      expect(getPathRegex(route).test("/post/a")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/b")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/b/c")).toEqual(true);
      expect(getPathRegex(route).test("/post/a/b/c/")).toEqual(true);
      expect(getPathRegex(route).test("/post")).toEqual(true);
      expect(getPathRegex(route).test("/post/")).toEqual(true);
    });

    it("extracts query parameters", () => {
      const route = "/post/[[...slug]]";
      expect(getPathRegex(route).exec("/post/a")?.groups).toEqual({
        slug: "/a",
      });
      expect(getPathRegex(route).exec("/post/a/")?.groups).toEqual({
        slug: "/a/",
      });
      expect(getPathRegex(route).exec("/post/a/b")?.groups).toEqual({
        slug: "/a/b",
      });
      expect(getPathRegex(route).exec("/post/a/b/c")?.groups).toEqual({
        slug: "/a/b/c",
      });
      expect(getPathRegex(route).exec("/post/a/b/c/")?.groups).toEqual({
        slug: "/a/b/c/",
      });
      expect(getPathRegex(route).exec("/post")?.groups).toEqual({
        slug: "",
      });
      expect(getPathRegex(route).exec("/post/")?.groups).toEqual({
        slug: "/",
      });
    });
  });

  describe("dynamic, catch all, optional catch all routes", () => {
    it("matches", () => {
      const route = "/foo/[foo]/bar/[...bar]/post/[[...slug]]";
      expect(getPathRegex(route).test("/foo/1/bar/1/2/post/a")).toEqual(true);
    });

    it("extracts query parameters", () => {
      const route = "/foo/[foo]/bar/[...bar]/post/[[...slug]]";
      expect(getPathRegex(route).exec("/foo/1/bar/1/2/post/a")?.groups).toEqual(
        {
          foo: "1",
          bar: "1/2",
          slug: "/a",
        }
      );
    });
  });
});

describe("getMatch", () => {
  it("generates route matcher", () => {
    const path = "/foo/[foo]/bar/[...bar]/post/[[...slug]]";
    const match = getMatch(path);
    expect(match("/foo/1/bar/1/2/post/a/b/c")).toEqual({
      match: true,
      params: {
        bar: ["1", "2"],
        foo: "1",
        slug: ["a", "b", "c"],
      },
    });
    expect(match("/foo/1/bar/1/2/post")).toEqual({
      match: true,
      params: {
        bar: ["1", "2"],
        foo: "1",
        slug: [],
      },
    });
    expect(match("/foo/1/bar/1/2/post/")).toEqual({
      match: true,
      params: {
        bar: ["1", "2"],
        foo: "1",
        slug: [],
      },
    });
    expect(match("/foo/")).toEqual({
      match: false,
      params: {},
    });
  });
});

describe("Router", () => {
  it(".match", () => {
    const filepath = "/foo/[foo].ts";
    const handler = jest.fn();
    const r = router({ [filepath]: handler });
    expect(r.match("/foo/1")).toEqual({
      handler,
      params: {
        foo: "1",
      },
      pathname: "/foo/[foo]",
    });
  });
});
