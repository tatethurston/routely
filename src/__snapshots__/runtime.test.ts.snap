// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`escapeRegExp escapes / 1`] = `"\\\\/"`;

exports[`escapeRegExp escapes /[foo] 1`] = `"\\\\/[foo]"`;

exports[`escapeRegExp escapes /[foo]/[bar]/[baz] 1`] = `"\\\\/[foo]\\\\/[bar]\\\\/[baz]"`;

exports[`escapeRegExp escapes /[foo]/bar/[baz] 1`] = `"\\\\/[foo]\\\\/bar\\\\/[baz]"`;

exports[`escapeRegExp escapes /[foo]/bar/[baz]/foo/[bar] 1`] = `"\\\\/[foo]\\\\/bar\\\\/[baz]\\\\/foo\\\\/[bar]"`;

exports[`escapeRegExp escapes /[foo]/baz 1`] = `"\\\\/[foo]\\\\/baz"`;

exports[`escapeRegExp escapes /_app 1`] = `"\\\\/_app"`;

exports[`escapeRegExp escapes /_debug 1`] = `"\\\\/_debug"`;

exports[`escapeRegExp escapes /_debug/health-check 1`] = `"\\\\/_debug\\\\/health-check"`;

exports[`escapeRegExp escapes /_document 1`] = `"\\\\/_document"`;

exports[`escapeRegExp escapes /_error 1`] = `"\\\\/_error"`;

exports[`escapeRegExp escapes /_error/ 1`] = `"\\\\/_error\\\\/"`;

exports[`escapeRegExp escapes /404 1`] = `"\\\\/404"`;

exports[`escapeRegExp escapes /api/[...segments] 1`] = `"\\\\/api\\\\/[...segments]"`;

exports[`escapeRegExp escapes /api/[[...segments]] 1`] = `"\\\\/api\\\\/[[...segments]]"`;

exports[`escapeRegExp escapes /api/bar 1`] = `"\\\\/api\\\\/bar"`;

exports[`escapeRegExp escapes /foo/[slug] 1`] = `"\\\\/foo\\\\/[slug]"`;

exports[`escapeRegExp escapes /not-found 1`] = `"\\\\/not-found"`;

exports[`escapeRegExp escapes /settings/ 1`] = `"\\\\/settings\\\\/"`;

exports[`escapeRegExp escapes /settings/bars/[bar] 1`] = `"\\\\/settings\\\\/bars\\\\/[bar]"`;

exports[`escapeRegExp escapes /settings/bars/[bar]/baz 1`] = `"\\\\/settings\\\\/bars\\\\/[bar]\\\\/baz"`;

exports[`escapeRegExp escapes /settings/foo 1`] = `"\\\\/settings\\\\/foo"`;

exports[`getPathRegex escapes / 1`] = `/\\^\\\\/\\\\/\\*\\$/`;

exports[`getPathRegex escapes /[foo] 1`] = `/\\^\\\\/\\(\\?<foo>\\[\\^/\\]\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /[foo]/[bar]/[baz] 1`] = `/\\^\\\\/\\(\\?<foo>\\[\\^/\\]\\+\\)\\\\/\\(\\?<bar>\\[\\^/\\]\\+\\)\\\\/\\(\\?<baz>\\[\\^/\\]\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /[foo]/bar/[baz] 1`] = `/\\^\\\\/\\(\\?<foo>\\[\\^/\\]\\+\\)\\\\/bar\\\\/\\(\\?<baz>\\[\\^/\\]\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /[foo]/bar/[baz]/foo/[bar] 1`] = `/\\^\\\\/\\(\\?<foo>\\[\\^/\\]\\+\\)\\\\/bar\\\\/\\(\\?<baz>\\[\\^/\\]\\+\\)\\\\/foo\\\\/\\(\\?<bar>\\[\\^/\\]\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /[foo]/baz 1`] = `/\\^\\\\/\\(\\?<foo>\\[\\^/\\]\\+\\)\\\\/baz\\\\/\\*\\$/`;

exports[`getPathRegex escapes /_app 1`] = `/\\^\\\\/_app\\\\/\\*\\$/`;

exports[`getPathRegex escapes /_debug 1`] = `/\\^\\\\/_debug\\\\/\\*\\$/`;

exports[`getPathRegex escapes /_debug/health-check 1`] = `/\\^\\\\/_debug\\\\/health-check\\\\/\\*\\$/`;

exports[`getPathRegex escapes /_document 1`] = `/\\^\\\\/_document\\\\/\\*\\$/`;

exports[`getPathRegex escapes /_error 1`] = `/\\^\\\\/_error\\\\/\\*\\$/`;

exports[`getPathRegex escapes /_error/ 1`] = `/\\^\\\\/_error\\\\/\\\\/\\*\\$/`;

exports[`getPathRegex escapes /404 1`] = `/\\^\\\\/404\\\\/\\*\\$/`;

exports[`getPathRegex escapes /api/[...segments] 1`] = `/\\^\\\\/api\\\\/\\(\\?<segments>\\.\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /api/[[...segments]] 1`] = `/\\^\\\\/api\\(\\?<segments>\\.\\*\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /api/bar 1`] = `/\\^\\\\/api\\\\/bar\\\\/\\*\\$/`;

exports[`getPathRegex escapes /foo/[slug] 1`] = `/\\^\\\\/foo\\\\/\\(\\?<slug>\\[\\^/\\]\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /not-found 1`] = `/\\^\\\\/not-found\\\\/\\*\\$/`;

exports[`getPathRegex escapes /settings/ 1`] = `/\\^\\\\/settings\\\\/\\\\/\\*\\$/`;

exports[`getPathRegex escapes /settings/bars/[bar] 1`] = `/\\^\\\\/settings\\\\/bars\\\\/\\(\\?<bar>\\[\\^/\\]\\+\\)\\\\/\\*\\$/`;

exports[`getPathRegex escapes /settings/bars/[bar]/baz 1`] = `/\\^\\\\/settings\\\\/bars\\\\/\\(\\?<bar>\\[\\^/\\]\\+\\)\\\\/baz\\\\/\\*\\$/`;

exports[`getPathRegex escapes /settings/foo 1`] = `/\\^\\\\/settings\\\\/foo\\\\/\\*\\$/`;
